### ISSUE TITLE
showtext_begin() now makes evaluate+knitr+rmarkdown think there is a plot to be replayed

### ISSUE THREAD

**Issue by cpsievert** [yixuan/showtext#41]
After #40, `evaluate::evaluate()` now returns a recordedplot when `new_device = FALSE`, even if the code shouldn't generate a plot. This presents some pretty major issues in knitr/rmarkdown.

```r
png()
dev.control(displaylist = "enable")
showtext::showtext_begin()
res <- evaluate::evaluate("1", new_device = FALSE)
res
#> [[1]]
#> $src
#> [1] "1"
#> 
#> attr(,"class")
#> [1] "source"
#> 
#> [[2]]
#> 
#> [[3]]
#> [1] "[1] 1\n"

class(res[[2]])
#> [1] "recordedplot"
```

**Comment by yixuan** [comment]
Yeah, in fact I have just discovered an issue with the `fig.showtext` option in knitr and was about to tell you that I need to delay the submission. 😄

**Comment by cpsievert** [comment]
Good to know, sorry for the oversight! 😅 

After researching this a bit more, it seems like a mistake for `showtext_begin()` to record by default, partially because [knitr wants to always call it (regardless of whether or not a plot is being generated)](https://github.com/yihui/knitr/blob/42b91f/R/block.R#L141-L146). 

I'm thinking `showtext_begin()` should gain a `record = FALSE` argument and `showtext_auto()` would register plot hooks that call `showtext_begin(record = TRUE)`, does that seem fool-proof to you?

**Comment by yixuan** [comment]
Can I ask a few questions to make sure I fully understand the current problem?

If I'm correct, the source of all these issues is that `showtext_auto()` does not affect `replayPlot()`, and the reason is that there is no hook available for `replayPlot()`, unlike `plot.new()`. To attach `showtext_begin()` to `replayPlot()`, we can use `recordGraphics()`, but that makes `showtext_begin()` always "create" a plot. So the paradox is that:
1. The current implementation is good for `knitr`, but insufficient for `shiny`.
2. `recordGraphics()` works for `shiny`, but causes problems to `knitr`.

Anyway, the core of the problem is to let `replayPlot()` be aware of `showtext`, so that `showtext_begin()` can be called every time `replayPlot()` is used. Am I correct?

**Comment by cpsievert** [comment]
> Anyway, the core of the problem is to let replayPlot() be aware of showtext, so that showtext_begin() can be called every time replayPlot() is used. Am I correct?

Yea, that's correct. Revisiting the example from #39 helps to point out the issue:

```r
library(showtext)
library(ggplot2)
font_add_google("Lobster", "lobster")
showtext_auto()
png("foo.png")
dev.control(displaylist = "enable")
qplot(1:10) + theme(text = element_text(family = "lobster"))
result <- recordPlot()
dev.off()
file.show("foo.png")

png("foo2.png")
replayPlot(result)
#> Error in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y,  : 
#>   polygon edge not found
#> In addition: Warning message:
#> In grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y,  :
#>   no font could be found for family "lobster"
```

My understanding is that this error occurs because `recordPlot()`+`replayPlot()` doesn't know to call `showtext_begin()` (for it's side-effects) prior to replaying the plot on a new device. It's worth noting, however, that if we call `plot.new()` prior to `replayPlot()`, this actually works (since `plot.new()` will call `showtext_begin()` when `showtext_auto()` is enabled):

```r
png("foo2.png")
plot.new()
replayPlot(result)
dev.off()
```

As it turns out, shiny actually calls `plot.new()` after opening a new device, so this the current implementation is a problem for shiny _assuming that `showtext_auto()` is globally enabled throughout the lifetime of the application_. However, this becomes an issue when using showtext in a local context, like this:

```r
library(shiny)
library(showtext)

with_showtext <- function(expr) {
  showtext_auto()
  on.exit(showtext_auto(FALSE))
  expr
}

ui <- fluidPage(plotOutput("p"))

server <- function(input, output, session) {
  output$p <- renderPlot({
    with_showtext(ggplot2::qplot(1:10))
  })
}

shinyApp(ui, server)
```

In this sort of scenario, when the plot is replayed (on resize), showtext is no longer enabled, so calling `plot.new()` isn't enough to make sure `showtext_begin()` is called before `replayPlot()`

**Comment by cpsievert** [comment]
Also, the reason why #40 ended up being problematic for evaluate+knitr is that [knitr open a new graphics device and calls `showtext_begin()` (if `fig.showtext = TRUE`) *for every code chunk*](https://github.com/yihui/knitr/blob/42b91f0/R/block.R#L141-L146) and thinks a plot has been generated whenever a call to `recordGraphics()` has been made during evaluation of that code block (and this is really a consequence of [the heuristics evaluate has to do guess when a plot has been made](https://github.com/yihui/evaluate/blob/42ebd9d/R/graphics.r#L9-L25)). That's why in #42, I've made `FALSE` the default for `record` in `showtext_begin()`, but I think it's safe for `showtext_auto()` to default to `TRUE` since, in that case, we know that when `showtext_begin()` is being called, it's because of `plot.new()`, and therefore we know a plot's being generated.

**Comment by yixuan** [comment]
Thanks for the detailed explanation!

If all we need is to let `replayPlot()` call `showtext_begin()` when needed, can we use the following hack? I bet you still remember the `trace()` trick.

```r
library(shiny)
library(ggplot2)
library(showtext)

with_showtext <- function(expr) {
    # These two lines are for the initial plot
    showtext_auto()
    on.exit(showtext_auto(FALSE))

    # The tracer is used for resizing
    trace(grDevices::replayPlot, tracer = quote(showtext_begin()), print = FALSE)
    on.exit(untrace(grDevices::replayPlot))

    expr
}

ui <- fluidPage(plotOutput("p"))

server <- function(input, output, session) {
    output$p <- renderPlot({
        with_showtext(ggplot2::qplot(1:10) + theme_bw(base_family = "serif"))
    })
}

shinyApp(ui, server)
```

**Comment by cpsievert** [comment]
#42 seems like a *much* better solution since `showtext_auto()` "just works" the way you'd expect it to. In fact, I was just able to verify that #42 actually makes the following work in RStudio Preview (v1.3.1073-1) without the `trace()` hack. This actually makes a lot of sense because I'm pretty sure the IDE uses `recordPlot()`/`replayPlot()` extensively.

```r
library(showtext)
font_add_google("Lobster", "lobster")
showtext_auto()
plot(1, pch = 16, cex = 3)
text(1, 1.1, "A fancy dot", family = "lobster", col = "steelblue", cex = 3)
```

**Comment by yixuan** [comment]
Great. I've merged your changes to the [record](https://github.com/yixuan/showtext/tree/record) branch for you to review. The main changes are:

1. The C function has an argument in the new version, and I have changed the call accordingly.
2. `record = FALSE` is default for backward compatibility.
3. In user API, the new `record` parameter is only for `showtext_auto()` but not for `showtext_begin()`. The reason is as follows.

Related to https://github.com/yixuan/showtext/commit/13e252cdf96438903e7ecd20fbef92601ee438d2, I guess `recordGraphics()` is only effective **after** `plot.new()` is called. To see this, try the following two code blocks:
```r
png("1.png")
dev.control(displaylist = "enable")
plot.new()
grDevices::recordGraphics(print(1), list(), .GlobalEnv)
res1 = recordPlot()
dev.off()

png("2.png")
dev.control(displaylist = "enable")
grDevices::recordGraphics(print(1), list(), .GlobalEnv)
plot.new()
res2 = recordPlot()
dev.off()

length(res1[[1]])
# [1] 3
length(res2[[1]])
# [1] 2
```

This explains why the previous `before.plot.new` hook does not work but the `plot.new` hook does. Then `showtext_begin(record = TRUE)` is only useful after `plot.new()` is called **AND** before any text rendering function is called. However, in most high-level plotting functions these two steps are encapsulated into a single call and it is hard to insert `showtext_begin` between the two steps. On the other hand, `showtext_auto()` relies on hooks, so it is able to insert `showtext_begin` at the proper location.

In a word, only add the `record` parameter to `showtext_auto()` and keep the API of `showtext_begin()` same as before.

Does this make sense to you?

**Comment by cpsievert** [comment]
> Then `showtext_begin(record = TRUE)` is only useful after `plot.new()` is called AND before any text rendering function is called.

Ah, yea, I agree then that it makes sense to not expose `record` in `showtext_begin()`. 

> In a word, only add the record parameter to showtext_auto() and keep the API of showtext_begin() same as before.

Yes, but why not default to `record = TRUE` (it would be awesome if `showtext_auto()` just worked in RStudio, etc)? I can't think of a scenario where `record = TRUE` isn't the behavior that you actually want.

Other than `record = FALSE`, the record branch looks good to me, but regardless of what the default of `record` ends up being, let's make sure to mention it in the NEWS. Also, if we go with `record = TRUE` (which I recommend), you could get rid of the RStudio known issue on the README.

As a side note, it'd be really great if showtext had some (visual) unit tests, perhaps using [vdiffr](https://github.com/r-lib/vdiffr). They can be a bit of a pain to maintain sometimes (I have several projects that use vdiffr), but I'd be happy to send you a PR to get you started and answer any questions you have.

**Comment by yixuan** [comment]
Great. Let's consider it in the next release, as CRAN will be offline soon and I also need to work on other projects at this moment.

**Comment by yixuan** [comment]
Hi @cpsievert, I think I should follow your advice and set `record = TRUE` by default. But did you get the Shiny app work? For some reason I still have issues.

**Comment by cpsievert** [comment]
Good to hear!

And sorry, the shiny example in https://github.com/yixuan/showtext/issues/41#issuecomment-670983568 was contrived and not fully though out...due to the way `renderPlot()` works, you'd actually need to `print()` the expression for the plot hooks to be called before they get removed (I don't think showtext necessarily needs to be concerned about that detail though)

```r
with_showtext <- function(expr) {
  showtext_auto()
  on.exit(showtext_auto(FALSE))
  print(expr)
}
```

In case you're interested, the use case I really care about is [`thematic`](https://github.com/rstudio/thematic) (specifically, `thematic_with_theme()`, which is similar in spirit to `with_showtext()`). [Thematic uses plot hooks as well](https://github.com/rstudio/thematic/blob/400fd3c8561d540ecd5c9d59610b50e32fd74cfc/R/hooks.R#L1-L6), and thanks to our back-and-forth, I realized I could "own" the enabling/recording myself by [calling `showtext_begin()`in a before new page hook](https://github.com/rstudio/thematic/blob/400fd3c8561d540ecd5c9d59610b50e32fd74cfc/R/hooks.R#L76-L79) and [record it in an after new page hook](https://github.com/rstudio/thematic/blob/400fd3c8561d540ecd5c9d59610b50e32fd74cfc/R/hooks.R#L31-L40).

**Comment by yixuan** [comment]
Great! I'll update some documentation and submit the current version to CRAN. Feel free to submit a new issue if you find anything wrong in the future.
